src/pylhe/__init__.py

```diff
@@ -120,14 +123,45 @@ def loads():
     pass
 
 
-def readLHEInit(thefile):
+def _extract_gzip_file(fileobj):
+    """
+    Checks to see if a file is compressed, and if so, extract it with gzip
+    so that the uncompressed file can be returned.
+
+    Args:
+        fileobj: A file object that can be a file or a compressed file.
+
+    Returns:
+        file: An uncompressed file.
+    """
+    if not Path(fileobj).name.lower().endswith(".gz"):
+        return fileobj
+
+    # Verify actually a gzip file
+    with open(fileobj, "rb") as gzip_file:
+        header = gzip_file.read(2)
+    gzip_magic_number = b"\x1f\x8b"
+    if header != gzip_magic_number:
+        raise OSError(
+            f"Input file {fileobj} is not a compressed file.\n"
+            + f"{fileobj} has header of {header} and not gzip's {gzip_magic_number}.\n"
+        )
+
+    with gzip.open(fileobj, "rb") as readfile:
+        extracted_file = readfile.read()
+
+    return extracted_file
```

Comment:
This is actually not good either. If you take the following madgraph script file:

```
$ cat drell-yan_test.mg5
generate p p > l+ l-
output drell-yan_test
launch drell-yan_test
```

and run it you'll have a file `drell-yan_test/Events/run_01/unweighted_events.lhe.gz`. If you take this script

```python
import gzip
from pathlib import Path

import pylhe

if __name__ == "__main__":
    unweighted_gzip = Path().cwd().joinpath("unweighted_events.lhe.gz")
    unweighted_file = Path().cwd().joinpath("unweighted_events.lhe")

    file_generator = pylhe.readLHE(unweighted_file)
    event = next(file_generator)
    print(f"uncompressed: {event.particles[-1].e}")

    file_generator = pylhe.readLHE(unweighted_gzip)
    event = next(file_generator)
    print(f"compressed:   {event.particles[-1].e}")
```

and run it you'll get

```
$ python issue_example.py 
uncompressed: 68.822088005
# and then a deluge of the contents of the LHE file being printed to screen
```

If these lines are replaced with

```python
    return gzip.open(fileobj, "rb")
```

then 

```
$ python example.py
uncompressed: 68.822088005
compressed:   68.822088005
```

So I probably need to look at what `ET.iterparse` is expecting if the goal is to have `_extract_gzip_file` return the same object type.

Reply:
[`xml.etree.ElementTree.iterparse()`](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.iterparse) want a file object to be able to iterate on, so returning a `gzip.GzipFile` like

```python
return gzip.open(fileobj, "rb")
```

is actually the right type to return. However, isn't it bad to be returning with an `open`? Ideally this would be done with a context manager, but it isn't clear to me how to properly copy all the necessary information as

```python
with gzip.open(fileobj, "rb") as gzip_file:
    extracted_fileobj = gzip.GzipFile(
        filename=gzip_file.filename, fileobj=gzip_file.read()
    )
return extracted_fileobj
```

runs into problems later in `xml.etree.ElementTree.iterparse`.

Reply:
...and it seems I was just making it hard for myself as

```python
with gzip.open(fileobj, "rb") as gzip_file:
    gzip_fileobj = gzip.GzipFile(filename=gzip_file.filename)
return gzip_fileobj
```

works.